### ISSUE TITLE
DataGrid: Empty Grid Placeholder is not identified by screen reader

### ISSUE THREAD

**Issue by kalpeshhpatel** [vmware/clarity#4090]
## Describe the bug

In Datagrid, when there are no records, placeholder appears. However, this placeholder (and it's state change) is not read by a screen-reader.

1.  Go to 'https://v2.clarity.design/datagrid/full
2.  Click on the 'Filter' button for the Name column.
3.  Filter name column with some random string, so that it returns an empty result.
4. Use `VoiceOver` or `NVDA` screen reader.
5.  Datagrid shows `No users found` message. However, this message is not readable by screen reader directly.

## Expected behavior

Whenever an empty grid placeholder appears, the screen reader should detect it and should be able to read.

## Versions

This issue is reproducible on the latest Clarity version.

**Comment by hippee-lee** [comment]
@chlane - Can you weigh in on the best and  practiced for a dynamic component? What is the best behavior users should expect here? I'm guessing that the placeholder needs to use the aria live service pattern to announce that its appeared. But that leads to questions about when the data is (re)populated, does that also need to be announced?

**Comment by chlane** [comment]
I'm thinking we need a separate element that is a polite live region that can be used to announce the status of a filtering operation. So when there are no results, it is populated with a message that conveys no results were found and when there are results the message conveys that results have been found.

**Comment by chlane** [comment]
To be clear, the text "No users found" in the element below, is readable by screen reader users when they move the reading cursor to it. 

`<div class="datagrid-placeholder datagrid-empty"><!----><div class="datagrid-placeholder-image"></div><!---->No users found<!----></div>`

But if it appears dynamically, blind screen reader users won't be aware of it at the moment it appears. This is why a status update using a live region make sense in this case.

**Comment by bdryanovski** [comment]
```ts
import { Component, ElementRef, PLATFORM_ID, Inject } from '@angular/core';
import { Items } from './providers/items';
import { AriaLiveService } from '../../utils/a11y/aria-live.service';
import { isPlatformBrowser } from '@angular/common';

@Component({
  selector: 'clr-dg-placeholder',
  providers: [AriaLiveService],
  template: `
        <div
            class="datagrid-placeholder"
            [class.datagrid-empty]="emptyDatagrid">
                <div class="datagrid-placeholder-image" *ngIf="emptyDatagrid"></div>
                <ng-content *ngIf="emptyDatagrid"></ng-content>
        </div>
    `,
  host: { '[class.datagrid-placeholder-container]': 'true' },
})
export class ClrDatagridPlaceholder<T = any> {
  constructor(
    private items: Items<T>,
    private el: ElementRef,
    private ariaLiveService: AriaLiveService,
    @Inject(PLATFORM_ID) private platformId: Object,
  ) {}

  ngAfterViewChecked(): void {
    if (isPlatformBrowser(this.platformId) && this.emptyDatagrid) {
      const element = this.el.nativeElement.querySelector('.datagrid-placeholder');
      if (element) {
        this.ariaLiveService.announce(element.innerText);
      }
    }
  }

  /**
   * Tests if the datagrid is empty, meaning it doesn't contain any items
   */
  public get emptyDatagrid() {
    return !this.items.loading && (!this.items.displayed || this.items.displayed.length === 0);
  }
}
```

With this little change, this could be done pretty easy. But I found a little issue with the AriaLive service and how SR is treating the content.

Let say that we have Datagrid with filter. By typing into the filter at some point we could get case that we don't have any result to show, so we will have to show the placeholder with the text. And that's fine for the first time. If we continue typing we will still continue making filter request and displaying the placeholder, we are still trying updating the `aria-live` container but the text that we are updating is the same. If in our test case it was "No users found" we will try to insert it again and again. Because the SR is not seeing any change it will skip it or this is the case for Chrome + VoiceoOver.

Is this ok or we have to announce it again until there is some result? For me, it seems normal that we will repeat the announcement that there are "No users found" until something change. 

If this is the case we have to modify the AriaLive to handle cases when the content is not changed but we still want to force the SR to pick it up. In that case, we could remove and recreate the hidden `aria-live` container (every container have unique ID so for the SR this will be a new message)

**Comment by hippee-lee** [comment]
IMO ( @clane is the expert here) I think it should announce again after the filter runs. e.g If I am typing into a filter and it announces no results after I type `xyz` It should continue to announce no results as I keep typing and there are still no results in the datagrid. 

What about using a `ViewChild ` instead of querying the DOM?

**Comment by hippee-lee** [comment]
IMO ( @clane is the expert here) I think it should announce again after the filter runs. e.g If I am typing into a filter and it announces no results after I type `xyz` It should continue to announce no results as I keep typing and there are still no results in the datagrid. 

What about using a `ViewChild ` instead of querying the DOM?

**Comment by bdryanovski** [comment]
It was a quick proof of concept code. It could be improved for sure. Also not sure about ViewChild will help in this case, but could try researching it a little more. But for sure think that `textContent` is the better property to use instead of `innerText`

And probably will need to add an interface to AriaLiveService to force update.

**Comment by chlane** [comment]
An update per keystroke could create a deep queue of redundant announcements that the user will have to listen to which could make the application unusable. Part of the reason for this is that screen readers use the same synthesized voice for the reading cursor as with live region announcements, which makes these two output streams indistinguishable.

Another consideration is that our applications are likely to have a lot of live region announcements coming from other components such as alerts. So in general live regions need to be used sparingly to let the user read/hear the content on the screen without too many interruptions. 

I recommend we either:

Implement live region updates on per keystroke basis but provide a checkbox to allow this feature to be disabled along with visually hidden text that explains the filtering behavior
 of this component to blind screen reader users so they know what is happening without hearing it real time

OR

Implement the live region updates to avoid redundancy and to happen as infrequently as possible or allow these decisions to be configured by the consuming application

**Comment by gnomeontherun** [comment]
Do we really need to implement a live region? The report is specifically about the placeholder text not being readable for some reason.

**Comment by chlane** [comment]
Agreed, dynamic data handling is out of the scope of this defect but we should address it. In my opinion it should be handled by the consuming application with a live region. Another option to deal with a large volume of status messages, would be to use visually hidden text to warn blind screen reader users that each keystroke will cause dynamic updates and that they should type slowly.

**Comment by chlane** [comment]
We should also aria-busy (https://www.w3.org/TR/wai-aria-1.1/#aria-busy) on the element with role="grid". This will indicate that the element is being modified and that assistive technologies MAY want to wait until the modifications are complete before exposing them to the user. The true/false value of aria-busy should be set according to the visibility of the "datagrid-spinner" element.

**Comment by bdryanovski** [comment]
So let start from somewhere, right now the placeholder don't have `aria-live` to announce the text and the focus is kept into the filter or any other component. So `aria-live` I think is the faster and simplest way to go.  

On the `aria-busy` there is no loading animation to listen for so this could be something hard to implement. We gonna need to link filters to the placeholder and know when we are applying filter and not just rendering empty Datagrid (empty Datagrid will be rendered with the same placeholder but I don't think `aria-busy` must be set to true)

If we implement it like that and don't make any changes to the AriaLiveService we won't get another announcement when the user change filter and he gets the same result (no data found).  If we want to announce it every time when he makes any interaction, we gonna need to update service to create/destroy the message so the SR will pick it up. And also maybe we will have to link the filter servies with the placeholder. Not sure for the last one, there could be something to use and don't create this link.

**Comment by gnomeontherun** [comment]
Let's backlog the aria-live and dynamic behaviors for another issue. Here we just want to make sure the content is readable. Let's fix that for this report and not get lost in the weeds.
